stdlib/public/Darwin/Compression/Compression.swift

```diff
@@ -202,12 +203,71 @@ public class OutputFilter {
 
 }
 
-@available(macOS 10.12, iOS 10.0, watchOS 3.0, tvOS 10.0, *)
-public class InputFilter {
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+public class InputFilter<D: DataProtocol> {
+
+  // Internal buffer to read bytes from a DataProtocol implementation
+  private class InputFilterBuffer<D: DataProtocol> {
+    private var _data: D // current input data
+    private var _remaining: Int // total bytes remaining to process in _data
+    private var _regionIndex: D.Regions.Index // region being read in _data
+    private var _regionRemaining: Int // remaining bytes to read in region being read in _data
+    public init(_ data: D) throws {
+      _data = data
+      _remaining = _data.count
+      _regionRemaining = 0
+      _regionIndex = _data.regions.startIndex
+      if _regionIndex != _data.regions.endIndex { _regionRemaining = _data.regions[_regionIndex].count }
+      // Advance to first non-zero region
+      try advance(by: 0)
+    }
+
+    // Return number of remaining bytes
+    public func remaining() -> Int { return _remaining }
+
+    // Call f with a buffer to the remaining bytes of the current contiguous region
+    public func withUnsafeBytes<R>(_ body: (UnsafeRawBufferPointer) throws -> R) rethrows -> R? {
+      if _remaining == 0 {
+        return try body(UnsafeRawBufferPointer(start: nil, count: 0))
+      } else {
+        let r = _data.regions[_regionIndex]
+        return try r.withUnsafeBytes { (region_buf: UnsafeRawBufferPointer) in
+          let src_buf = UnsafeRawBufferPointer(
+            start: region_buf.baseAddress!.assumingMemoryBound(to: UInt8.self) + region_buf.count - _regionRemaining,
+            count: _regionRemaining)
+          return try body(src_buf)
+        }
+      }
+    }
+
+    // Consume n bytes in the current region (n can be 0, up to _regionRemaining), and move to next non-empty region if needed
+    // post-condition: _remaining == 0 || _regionRemaining > 0
+    public func advance(by n: Int) throws {
+
+      // Sanity checks
+      if n > _regionRemaining { throw FilterError.invalidState } // invalid n
+
+      // Update counters
+      _regionRemaining -= n
+      _remaining -= n
+
+      // Move to next non-empty region if we are done with the current one
+      let r = _data.regions
+      while _regionRemaining == 0 {
+        _regionIndex = r.index(after: _regionIndex)
```

Comment:
It may be a good idea to use a mutating version here.
```swift
r.formIndex(after: &_regionIndex)
```

Reply:
ok